### PR TITLE
Update UISearchDisplayController to new UISearchController

### LINKED ISSUE #414: UISearchDisplayController has been deprecated in iOS 8
The UserVoice SDK uses the `UISearchDisplayController` which has been deprecated as of iOS 8.  This can be verified in Apple's documentation:

> Important: UISearchDisplayController is deprecated in iOS 8. (Note that UISearchDisplayDelegate is also deprecated.) To manage the presentation of a search bar and display search results in iOS 8 and later, instead use UISearchController.

See: https://developer.apple.com/library/ios/documentation/UIKit/Reference/UISearchDisplayController_Class/

This is not only adding compiler warnings, but introduces the risk that with upcoming releases of iOS, this project will not compile at all.

See the following screenshot as an example:
screen shot 2016-05-11 at 9 47 02 am

### PR DESCRIPTION
**Pull request by donnywdavis**
Resolves #414 to update the deprecated UISearchDisplayController to the current UISearchController to be compatible with iOS 9+.

### CODE REVIEW

File: Classes/UVSuggestionSearchResultsController.m [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    status.text = [suggestion.status uppercaseString];
+}
+
+#pragma mark - Table view data source
+
+- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
+    tableView.backgroundView = nil;
+    tableView.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
+    
+    if (_searchResults.count == 0) {
```
Comment (austintaylor) [inline Classes/UVSuggestionSearchResultsController.m]: This isn't really the right place to do this. I would think this controller should have like an `updateResults` method that could do this part, and also call `[self.tableView reloadData]`.

File: Classes/UVSuggestionSearchResultsController.m [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+- (void)initCellForResult:(UITableViewCell *)cell indexPath:(NSIndexPath *)indexPath {
+    [self initCellForSuggestion:cell indexPath:indexPath];
+}
+
+- (void)customizeCellForResult:(UITableViewCell *)cell indexPath:(NSIndexPath *)indexPath {
+    [self customizeCellForSuggestion:[_searchResults objectAtIndex:indexPath.row] cell:cell];
+}
+
+- (void)initCellForSuggestion:(UITableViewCell *)cell indexPath:(NSIndexPath *)indexPath {
```
Comment (austintaylor) [inline Classes/UVSuggestionSearchResultsController.m]: It appears that this method and the one below it are copied verbatim from UVSuggestionListViewController. Could you extract them into a helper so that the function bodies are not duplicated?

File: Classes/UVWelcomeSearchResultsController.m [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    [_instantAnswerManager customizeCell:cell forSuggestion:(UVSuggestion *)model];
+}
+
+#pragma mark - Table view data source
+
+- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
+    tableView.backgroundView = nil;
+    tableView.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
+    
+    if (_searchResults.count == 0) {
```
Comment (austintaylor) [inline Classes/UVWelcomeSearchResultsController.m]: Same comment as above

File: Classes/UVWelcomeViewController.m [hunk 4]
```diff
@@ -280,60 +248,44 @@ - (void)logoTapped {
...
+
+// DDSearch
+- (void)updateSearchResultsForSearchController:(UISearchController *)searchController {
+    _instantAnswerManager.searchText = searchController.searchBar.text;
+    [_instantAnswerManager search];
+    
+    if (_searchController.searchResultsController) {
+        UVWelcomeSearchResultsController *searchResultsTVC = (UVWelcomeSearchResultsController *)_searchController.searchResultsController;
+        searchResultsTVC.searchResults = self.searchResults;
+        [searchResultsTVC.tableView reloadData];
```
Comment (austintaylor) [inline Classes/UVWelcomeViewController.m]: This is where I would expect an `updateResults` method to be called, as suggested above.

File: Classes/UVSuggestionListViewController.m [hunk 5]
```diff
@@ -231,59 +222,52 @@ - (void)showSuggestion:(UVSuggestion *)suggestion {
...
+#pragma mark ==== UISearchResultsUpdating Methods ====
+
+// DDSearch
+- (void)updateSearchResultsForSearchController:(UISearchController *)searchController {
+    [UVSuggestion searchWithForum:_forum query:searchController.searchBar.text delegate:self];
+    
+    if (_searchController.searchResultsController) {
+        UVSuggestionSearchResultsController *searchResultsTVC = (UVSuggestionSearchResultsController *)_searchController.searchResultsController;
+        searchResultsTVC.searchResults = self.searchResults;
+        [searchResultsTVC.tableView reloadData];
```
Comment (austintaylor) [inline Classes/UVSuggestionListViewController.m]: Also here (see comments below)

File: Classes/UVSuggestionListViewController.m [hunk 6]
```diff
@@ -77,11 +79,13 @@ - (void)didSearchSuggestions:(NSArray *)theSuggestions {
     for (UVSuggestion *suggestion in theSuggestions) {
         [ids addObject:[NSNumber numberWithInteger:suggestion.suggestionId]];
     }
-    [UVBabayaga track:SEARCH_IDEAS searchText:_searchBar.text ids:ids];
-    if (FORMSHEET) {
-        [_tableView reloadData];
+    // DDSearch
```
Comment (austintaylor) [inline Classes/UVSuggestionListViewController.m]: All DDSearch comments should be removed.

File: Classes/UVSuggestionSearchResultsController.m (hunk 2 above)
Comment (austintaylor) [inline Classes/UVSuggestionSearchResultsController.m]: Or they could be pulled up into the UVBaseController if that doesn't cause any issues.

**Comment by donnywdavis** [comment]
Thanks @austintaylor. I thought I had removed all of the DDSearch comments but looks like I missed some. I will take care of the issues you mentioned and resubmit.

File: Classes/UVWelcomeViewController.m [hunk 7]
```diff
@@ -280,60 +248,44 @@ - (void)logoTapped {
...
-        [_tableView reloadData];
+#pragma mark ==== UISearchResultsUpdating Methods ====
+
+// DDSearch
+- (void)updateSearchResultsForSearchController:(UISearchController *)searchController {
+    _instantAnswerManager.searchText = searchController.searchBar.text;
+    [_instantAnswerManager search];
+    
+    if (_searchController.searchResultsController) {
+        UVWelcomeSearchResultsController *searchResultsTVC = (UVWelcomeSearchResultsController *)_searchController.searchResultsController;
```
Comment (donnywdavis) [inline Classes/UVWelcomeViewController.m]: @austintaylor The reason that I had the call to `updateSearchResultsForSearchController:` in `didUpdateInstantAnswers` is b/c it was never displaying the results since the results were being returned after the search results controller was initialized and attempted to load. 

A fix that I have that seems to be working was to move the code for setting up the search results controller out of `updateSearchResultsForSearchController:` and into `didUpdateInstantAnswers`. The `updateSearchResultsForSearchController:` method still grabs the search text and calls `[_instantAnswerManager search];` and that is it.

`#pragma mark ==== UISearchResultsUpdating Methods ====

- (void)updateSearchResultsForSearchController:(UISearchController *)searchController {
    _instantAnswerManager.searchText = searchController.searchBar.text;
    [_instantAnswerManager search];
}

#pragma mark ===== Search handling =====

- (void)didUpdateInstantAnswers {
    if (_searchController.active && ![_searchController.searchBar.text isEqualToString:@""]) {
        if (_searchController.searchResultsController) {
            UVWelcomeSearchResultsController *searchResultsTVC = (UVWelcomeSearchResultsController *)_searchController.searchResultsController;
            searchResultsTVC.searchResults = self.searchResults;
            searchResultsTVC.tableView.backgroundView = nil;
            searchResultsTVC.tableView.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
            [searchResultsTVC.tableView reloadData];
        }
    }
}`

I then changed the `searchBar:selectedScopeButtonIndexDidChange:` method to call `[self didUpdateInstantAnswers];` instead of `updateSearchResultsForSearchController:` to update the results when the scope is changed.

Hopefully this helps make it a little easier to follow the code. Any other recommendations on this piece?

Comment (austintaylor) [inline Classes/UVWelcomeViewController.m]: Makes sense. Works for me.

**Comment by donnywdavis** [comment]
@austintaylor I made the changes you suggested and pushed them up. They should be available for you to take a look now.

File: Classes/UVSuggestionListViewController.m [hunk 8]
```diff
@@ -346,19 +269,29 @@ - (void)initNavigationItem {
     }
 }
 
+- (void)dismiss {
```
Comment (austintaylor) [inline Classes/UVSuggestionListViewController.m]: @donnywdavis I just noticed that if you launch the forum directly it will not dismiss because you are overriding this method without calling super. Why is this needed here when we are already cleaning things up in dealloc?

Comment (donnywdavis) [inline Classes/UVSuggestionListViewController.m]: When I initially started making the changes for the UISearchController I had to add that method b/c it would not dismiss unless I set the searchResultsController to nil when it was called. This was probably due to having something coded wrong initially. I removed that method call now after making all of the recent changes and it seems to behave as expected without it. I pushed the change up just now.
